### PR TITLE
jyg-UID2-218-Google-esp-integration

### PR DESCRIPTION
**Pull request by jingyi-gao-ttd**

### CODE REVIEW

File: src/uid2Sdk.ts [hunk 1]
```diff
@@ -111,6 +124,11 @@ export class UID2 {
         const identity = this._opts.identity ? this._opts.identity : this._cookieManager.loadIdentityFromCookie();
         const validatedIdentity = this.validateAndSetIdentity(identity);
         if (validatedIdentity) this.triggerRefreshOrSetTimer(validatedIdentity);
+        if (this._opts.enableESP && !this._espEnabled) {
```
Comment (atarassov-ttd) [inline src/uid2Sdk.ts:127]: Given that google has renamed ESP to SecureSignals, I think we should follow suit and stop using "ESP" in our codebase too.

File: src/uid2Sdk.ts [hunk 2]
```diff
@@ -26,11 +26,20 @@ export class UID2 {
     }
     static IdentityStatus = IdentityStatus;
     static EventType = EventType;
-    static setupGoogleTag = setupGoogleTag;
+    static setupGoogleTag() {
```
Comment (atarassov-ttd) [inline src/uid2Sdk.ts:29]: Should this be renamed to something like `setupGoogleSecureSignals`?

File: src/uid2Esp.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+    }
+    if (this.isUID2SDKIntegrated() && 'getAdvertisingToken' in window.__uid2!) {
+      return window.__uid2!.getAdvertisingToken!.bind(window.__uid2);
+    }
+  };
+}
+
+declare global {
+  interface Window {
+    __uid2Esp: UID2EncryptedSignalProvider;
```
Comment (atarassov-ttd) [inline src/uid2Esp.ts:54]: General style question: `UID2EncryptedSignalProvider` or `Uid2EncryptedSignalProvider`?

File: examples/google-esp-integration/server_only/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Server-Only UID2 Integration Example
+
+This example demonstrates how a content publisher can use the UID2 services to implement the [custom UID2 integration workflow](https://github.com/UnifiedID2/uid2docs/blob/main/api/v1/guides/custom-publisher-integration.md).
+
+For an example application using the [Client-Side Identity JavaScript SDK](https://github.com/UnifiedID2/uid2docs/blob/main/api/v1/sdks/client-side-identity-v1.md), see [UID2 SDK Integration Example](../standard/README.md).
```
Comment (atarassov-ttd) [inline examples/google-esp-integration/server_only/README.md:5]: Is there an ESP integration example which uses JS SDK?

File: examples/google-esp-integration/server_only/views/header.html [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+      const uid2Token = getCookie('identity')
+      if (uid2Token === "") return
+      return JSON.parse(decodeURIComponent(uid2Token)).advertising_token
+    }
+
+    window.getAdvertisingToken = getAdvertisingToken;
+  
+    $(document).ready(function() {
+      if(window.location.href.endsWith('/login')){
+        window.googletag.cmd.push(window.googletag.encryptedSignalProviders.clearAllCache)
```
Comment (atarassov-ttd) [inline examples/google-esp-integration/server_only/views/header.html:37]: I think `encryptedSignalProviders` is deprecated -- should we switch to `secureSignalProviders` everywhere?

File: examples/google-esp-integration/server_only/README.md (hunk 4 above)
Comment (jingyi-gao-ttd) [reply examples/google-esp-integration/server_only/README.md:5]: Yes, but only with SDK v3

File: src/uid2Esp.ts [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    getAdvertisingToken?: () => string;
```
[the remark this answers is not in the document]
Comment (jingyi-gao-ttd) [reply src/uid2Esp.ts:55]: the function can return null or undefined, as we checked uid2 existence before we register the handler, so when this function get called by GPT, it should never return null or undefined

[the remark this answers is not in the document]
Comment (jingyi-gao-ttd) [reply src/uid2Esp.ts:55]: > Can `getAdvertisingToken()` function return `null` or `undefined`?

Good point, the function can return null or undefined, as we checked uid2 existence before we register the handler, so when this function get called by GPT, it should never return null or undefined

File: src/uid2Esp.ts (hunk 3 above)
Comment (jingyi-gao-ttd) [reply src/uid2Esp.ts:54]: We have both `UID2xxxx` and `Uid2xxx` in SDK 3, I was not sure which name convention to follow.  I will change it to `Uid2EncryptedSignalProvider`

File: src/uid2Sdk.ts (hunk 2 above)
Comment (jingyi-gao-ttd) [reply src/uid2Sdk.ts:29]: This function exists in old sdks, it was updated for backward compatibility, I think we'd better not rename it.

File: examples/google-esp-integration/server_only/README.md (hunk 4 above)
Comment (jingyi-gao-ttd) [reply examples/google-esp-integration/server_only/README.md:5]: All docs are updated in another branch, which I will update them again once this PR get merged

File: src/uid2Sdk.ts (hunk 2 above)
Comment (lionell-pack-ttd) [reply src/uid2Sdk.ts:29]: Maybe it's a good opportunity to provide a new name and also support the old one - that sets us on the path to eventually deprecate the old name.

File: src/uid2SecureSignal.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+    if (uid2Handler()) {
+      window.googletag = window.googletag || {
+        cmd: [],
+      };
+      window.googletag.secureSignalProviders =
+        window.googletag.secureSignalProviders || [];
+      window.googletag.secureSignalProviders.push({
+        id: "uidapi.com",
+        collectorFunction: () => Promise.resolve(uid2Handler()),
```
Comment (lionell-pack-ttd) [inline src/uid2SecureSignal.ts:29]: Is the goal to call `uid2Handler()` twice - once to make sure it returns something truthy, and also later when Secure Signals invokes the collector function?

File: src/uid2Esp.ts (hunk 3 above)
Comment (lionell-pack-ttd) [reply src/uid2Esp.ts]: Yeah, I've noticed that inconsistency too - I'd like to fix up as many as possible.

File: src/uid2Esp.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (jingyi-gao-ttd) [reply src/uid2Esp.ts]: This function could be defined by the publisher and be called in the ESP script, I am not sure if there is a better to way to achieve this?
This has been named `getUid2AdvertisingToken`, would it be fine like this? or we can name it to `__uid2__getUid2AdvertisingToken` ?

File: src/uid2SecureSignal.ts (hunk 7 above)
Comment (jingyi-gao-ttd) [reply src/uid2SecureSignal.ts:29]: Yes, as when `secureSignalProviders.push` is called, the secureSignalProviders will immediately invoke the collector function. The first call could avoid this push get triggered when advertisting token is not ready and cache a invalid value such as undefined or null

File: src/integrationTests/uid2SecureSignalHandler.test.ts [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+    
+  beforeEach(() => {
+    uid2 = new UID2();
+    sdkWindow.__uid2Esp = new Uid2SecureSignalProvider();
+    sendSignalMock = jest.fn();
+  });
+  
+  describe("when initialize UID2 with enableSecureSignals set to true", () => {
+    it("should not fail if uid2ESP script has not loaded", () => {
+      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
```
Comment (lionell-pack-ttd) [inline src/integrationTests/uid2SecureSignalHandler.test.ts:46]: I think it's OK to disable `ban-ts-comment` for the entire file in tests.

File: src/integrationTests/uid2SecureSignaltest.ts [hunk 9]
```diff
@@ -0,0 +1,192 @@
...
+    //@ts-ignore
+    window.__uid2Esp = undefined
+  })
+        
+  describe("when SDK enable esp with identity", () => {
+    describe("When script loaded before SDK loaded", () => {
+      test("should send signal to Google ESP when SDK initialized", async () => {
+        uid2.init({ identity, enableSecureSignals: true });
+        //@ts-ignore
+        expect(uid2ESP.retrieveAdvertisingTokenHandler()!()).toBe(identity.advertising_token);
```
Comment (lionell-pack-ttd) [inline src/integrationTests/uid2SecureSignaltest.ts:134]: It would be better to make this check the identity the push mock is called with. This `expect` tests a private method, so the test might break if you do an internal refactor, and it also doesn't test that the token is actually passed to Secure Signals.

Comment (jingyi-gao-ttd) [reply src/integrationTests/uid2SecureSignaltest.ts:134]: Yes, this line here is to test the `retrieveAdvertisingTokenHandler` is calling the correct function, in this case, it should use `__uid2.getAdverstingToken` as `retrieveAdvertisingTokenHandler`
At line 141 `expect(secureSignalProvidersResolveMock).toHaveBeenCalledWith(identity.advertising_token)` is checking if the token is passed to SecureSignal as the push mock is defined as this  `secureSignalProvidersPushMock = jest.fn(async (p: EncryptedSignalProvider) => {
    secureSignalProvidersResolveMock(await p.collectorFunction())
  })`

**Comment by lionell-pack-ttd** [comment]
I'm thinking through a few scenarios, and there's one which I think we should consider:

- The user visits the publisher at time T. We save a token which expires at (T+24 hours) and Secure Signals caches it.
- The user revisits the site at (T+2 hours), and we refresh the token - it now expires at (T+26 hours), but Secure Signals ignores it.
- The user revisits the site at (T+25 hours). SS's cache is over 24 hours old.
-- At this point, the SDK will return a token (which expires at T+26 hours), but also fire off a refresh request.
-- If SS tries to refresh its cache too soon, it will get the token which expires at T+26 hours.
-- Very soon after, the new token is available, but we can't update SS until (T+49 hours), so the token is expired for most of that time.

If we checked whether the SDK is trying to refresh the token, and wait until the new value is available, we could prevent that situation. Or maybe it should depend on how long until expiry - if it's 23 hours 40 seconds until expiry, still use it - if it's 10 minutes until expiry, wait for the refresh?

File: src/uid2SecureSignal.ts [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+
+declare global {
+  interface Window {
+    __uid2Esp: Uid2SecureSignalProvider;
+    getUid2AdvertisingToken?: () => Promise<string | null | undefined>;
+  }
+}
+
+(function () {
+  window.__uid2Esp = new Uid2SecureSignalProvider();
```
Comment (atarassov-ttd) [inline src/uid2SecureSignal.ts:80]: Should this be renamed to `__uid2SecureSignalProvider` or something like that?

File: src/uid2SecureSignal.ts [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+    ) {
+      console.warn("Please implement `getUid2AdvertisingToken`");
+      return;
+    }
+
+    try {
+      const token = await window.getUid2AdvertisingToken!();
+      this.registerSecureSignalProvider(token);
+    } catch (e) {
+      console.error(e);
```
Comment (atarassov-ttd) [inline src/uid2SecureSignal.ts:39]: Is it a good practice to log to console on errors?

Comment (atarassov-ttd) [reply src/uid2SecureSignal.ts:39]: Should we be providing more context than just dump of exception?

File: src/integrationTests/async.spec.ts [hunk 12]
```diff
@@ -175,3 +195,165 @@ describe('when window.__uid2.init is called on SdkLoaded from a callback', () =>
     expect(sdkWindow.__uid2.getAdvertisingToken()).toBe(identity.advertising_token);
   });
 });
+
+describe('when getFreshAdvertisingToken is called before init complete', () => {
+  describe('when initialising with a valid identity', () => {
+    const identity = makeIdentity();
+    test('it should resolve promise after invoking the callback', () => {
```
Comment (lionell-pack-ttd) [inline src/integrationTests/async.spec.ts:202]: Why does it matter that the getFreshAdvertisingToken promise is resolved after calling the init callback?

File: src/integrationTests/async.spec.ts [hunk 13]
```diff
@@ -175,3 +195,165 @@ describe('when window.__uid2.init is called on SdkLoaded from a callback', () =>
...
+        return token;
+      });
+      uid2.init({ callback: callback, identity: identity });
+      jest.runAllTimers();
+      return expect(p).resolves.toBe(identity.advertising_token);
+    });
+  });
+
+  describe('when initialising with an invalid identity', () => {
+    test('it should reject promise after invoking the callback', () => {
```
Comment (lionell-pack-ttd) [inline src/integrationTests/async.spec.ts:214]: Again here - I'm not sure why the init callback matters if you're testing getFreshAdvertisingToken.

File: src/integrationTests/async.spec.ts [hunk 14]
```diff
@@ -175,3 +195,165 @@ describe('when window.__uid2.init is called on SdkLoaded from a callback', () =>
...
+  describe('when initalising with a non-expired identity which requires a refresh', () => {
+    test('it should resolve with new advertising token after refresh', () => {
+      const originalIdentity = makeIdentity({
+        refresh_from: Date.now() - 100000,
+      });
+      const updatedIdentity = makeIdentity({
+        advertising_token: 'updated_advertising_token'
+      });
+      const p = uid2.getFreshAdvertisingToken().then((token: any) => {
+        expect(callback).toHaveBeenCalled();
```
Comment (lionell-pack-ttd) [inline src/integrationTests/async.spec.ts:233]: Callback doesn't seem to have anything to do with this test. I think you can achieve the same result just with `const p = uid2.getFreshAdvertisingToken()`;

File: src/integrationTests/async.spec.ts [hunk 15]
```diff
@@ -175,3 +195,165 @@ describe('when window.__uid2.init is called on SdkLoaded from a callback', () =>
...
+      });
+      uid2.init({ callback: callback, identity: originalIdentity });
+      xhrMock.responseText = JSON.stringify({ status: 'error' });
+      xhrMock.onreadystatechange(new Event(''));
+      return expect(uid2.getFreshAdvertisingToken()).rejects.toBeInstanceOf(Error);
+    });
+  });
+
+  describe('when disconnect() has been called', () => {
+    test('it should reject promise', () => {
```
Comment (lionell-pack-ttd) [inline src/integrationTests/async.spec.ts:337]: A lot of these tests are really just testing the same thing as the `getAdvertisingTokenAsync` tests - that's a lot of extra maintenance and code.

File: src/uid2ApiClient.ts [hunk 16]
```diff
@@ -1,122 +1,189 @@
...
-        const arrayBuffer = new Uint8Array(text.length);
-        for (let i = 0; i < text.length; i++) {
-            arrayBuffer[i] = text.charCodeAt(i);
+  public getFreshAdvertisingToken() {
+    return new Promise<string>((resolve, reject) => {
+      this._refreshingPromise = { resolve: resolve, reject: reject };
+    });
+  }
+
+  private _handleEvent(eventType: EventType, payload: Uid2CallbackPayload) {
```
Comment (lionell-pack-ttd) [inline src/uid2ApiClient.ts:77]: There should be a simpler way to do this - I don't like introducing this extra responsibility here. The SDK client is really just that - a client. It doesn't know about the SDK, or integrations, and it's likely to be used in places where we aren't using that (e.g. Prebid.js).

The existing data flow for renewing a token looks roughly like this:
image

With this change, there's an extra loop back down to this class after it's been through the event system:
image

You could move all of these changes somewhere else easily - the only thing they need to know about the API client is whether there's an active request, and you can just use `.hasActiveRequests()` to check that. That way you can avoid the extra dependency here.

File: src/integrationTests/async.spec.ts [hunk 17]
```diff
@@ -49,8 +56,29 @@ describe('when getAdvertisingTokenAsync is called before init', () => {
...
+  describe("when initalising with a non-expired identity which requires a refresh", () => {
+    test("it should resolve updated advertising", () => {
+      const originalIdentity = makeIdentity({
+        refresh_from: Date.now() - 100000,
+      });
+      const updatedIdentity = makeIdentity({
+        advertising_token: "updated_advertising_token",
+      });
+      const p = uid2.getAdvertisingTokenAsync().then((token: any) => {
+        expect(callback).toHaveBeenCalled();
```
Comment (lionell-pack-ttd) [inline src/integrationTests/async.spec.ts:68]: I don't think anything to do with callback belongs in this test - it's not relevant.

File: src/integrationTests/secureSignal.test.ts [hunk 18]
```diff
@@ -0,0 +1,274 @@
+// Copyright (c) 2021 The Trade Desk, Inc
```
Comment (lionell-pack-ttd) [inline src/integrationTests/secureSignal.test.ts:1]: You can remove the copyright notice - we just keep the one version for the whole repo now.

File: src/uid2ApiClient.ts (hunk 16 above)
Comment (jingyi-gao-ttd) [reply src/uid2ApiClient.ts]: I've now put those changes into uid2PromiseHandler, which is used to take care of the `getAdvertisingTokenAsync`, as we now change the behavior of `getAdvertisingTokenAsync`, I think it make sense to still let uid2PromiseHandler to hold those logic

File: src/integrationTests/secureSignal.test.ts [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+describe("when use script without SDK integrated", () => {
+  window.__uid2 = undefined;
+
+  describe("when getUid2AdvertisingToken exists and returns valid advertisingToken", () => {
+    test("should send signal to Google ESP", async () => {
+      window.getUid2AdvertisingToken = getAdvertisingTokenMock;
+      getAdvertisingTokenMock.mockReturnValue(Promise.resolve("testToken"));
+      uid2ESP = new Uid2SecureSignalProvider();
+      //@ts-ignore
+      expect(await uid2ESP.retrieveAdvertisingTokenHandler()!()).toBe(
```
[the remark this answers is not in the document]
Comment (jingyi-gao-ttd) [reply src/integrationTests/secureSignal.test.ts]: I was trying to test if the retrieveAdvertisingTokenHandler can differentiate UID2 SDK and server-only and return the correct function, I agree it is redundant as we tested the return value already. I've removed those from tests

File: src/secureSignal.ts [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        .catch((error: any) => {
+          if (attempts >= retries) {
+            reject(error);
+          } else {
+            attempt();
+          }
+        });
+    }
+
+    while (attempts < retries) {
```
Comment (lionell-pack-ttd) [inline src/secureSignal.ts:82]: `attempt()` creates a promise, but doesn't wait for it - so it will create `retries` promises immediately, and race them - whichever one either rejects or resolves first will win. So if one fails immediately, it will reject - even if the rest would have succeeded and returned a valid token.

File: src/uid2Sdk.ts [hunk 21]
```diff
@@ -1,230 +1,354 @@
...
-    static get COOKIE_NAME() {
-        return "__uid_2";
+  static get VERSION() {
+    return "3.0.0";
+  }
+  static get COOKIE_NAME() {
+    return "__uid_2";
+  }
+  static get DEFAULT_REFRESH_RETRY_PERIOD_MS() {
+    return 2000;
```
Comment (lionell-pack-ttd) [inline src/uid2Sdk.ts:27]: Why the change to 2000? I'm not sure where the original number came from, and 2000 seems sensible to me, but I'm curious.

File: src/uid2Sdk.ts [hunk 22]
```diff
@@ -1,230 +1,354 @@
...
+  public init(opts: Uid2Options) {
+    this.initInternal(opts);
+  }
+  public getAdvertisingToken() {
+    return this.getIdentity()?.advertising_token ?? undefined;
+  }
+  public setIdentity(identity: Uid2Identity) {
+    const validatedIdentity = this.validateAndSetIdentity(identity);
+    if (validatedIdentity) {
+      this.triggerRefreshOrSetTimer(validatedIdentity);
```
Comment (lionell-pack-ttd) [inline src/uid2Sdk.ts:86]: There might already be a refresh timer running when this is called. Rather than cancelling it here, it's worth thinking about whether the `setRefreshTimer` method should cancel any currently active timer - it doesn't make sense to have 2 of them active.

File: src/uid2Sdk.ts (hunk 21 above)
Comment (jingyi-gao-ttd) [reply src/uid2Sdk.ts:27]: There was a conversation here: https://thetradedesk.slack.com/archives/C045X131YD9/p1670212126775819 
but I just realise we didn't put the timeout on the request itself, thus the getAdvertisingTokenAsync will still be hanging by the request, I will add timeout to the api client as well

File: src/uid2Sdk.ts (hunk 22 above)
Comment (lionell-pack-ttd) [reply src/uid2Sdk.ts:86]: I think there's another scenario that needs to be addressed.

1. SDK is initialized and loads an old identity from the cookie. It needs to be refreshed so it sends a refresh request.
2. The publisher sets a new identity which doesn't need to be refreshed. The timer starts to check again in 5s.
3. The refresh request from step 1 completes and the refreshed identity from the API call (based on the cookie) is set, so the identity the publisher set in step 2 is lost.
4. When the timer triggers, it will see the refreshed identity, not the new one provided in the call to setIdentity.

File: src/integrationTests/autoRefresh.test.ts [hunk 23]
```diff
@@ -48,11 +53,17 @@ describe("when auto refreshing a non-expired identity which does not require a r
   });
   test("should set refresh timer", () => {
     expect(setTimeout).toHaveBeenCalledTimes(1);
-    expect(clearTimeout).not.toHaveBeenCalled();
+    expect(clearTimeout).toHaveBeenCalledTimes(1);
```
Comment (lionell-pack-ttd) [inline src/integrationTests/autoRefresh.test.ts:56]: This doesn't seem right - before it was saying that the timeout should be set and not cancelled (i.e. we know it will run). Now it's saying the timeout should be set and cancelled. I actually think this is uncovering a bug in the `beforeEach` method, and one of the other tests was always wrong (the "should not invoke the callback" test just above this - it should invoke the callback because it's got a valid identity the callback could use). Let me know if you want to pair on this one.

File: src/uid2Sdk.ts (hunk 22 above)
Comment (jingyi-gao-ttd) [reply src/uid2Sdk.ts:86]: In this case, the refreshed token from step 1 is the most recent identity, shouldn't we use it? The case that is problematic would be if the identity from the cookie and from the publisher is a different user? but this shouldn't happen if publisher use SDK correctly?

File: src/integrationTests/autoRefresh.test.ts (hunk 23 above)
Comment (jingyi-gao-ttd) [reply src/integrationTests/autoRefresh.test.ts:56]: I found the cause of this issue is that `this._refreshTimerId` is not get reset after the function gets executed, I've now add `this._refreshTimerId = null;` to the end of the setTimeout function

Comment (lionell-pack-ttd) [reply src/integrationTests/autoRefresh.test.ts:56]: Good catch - that fixes a possible bug. But this test looks wrong - after you init with a token which doesn't need to be refreshed, you expect setTimeout to be called once, but clearTimeout shouldn't be called. Or if the timer has already run once, and it's set a new timer, then setTimeout should have been called twice, but there's still no reason it should call clearTimeout.

So why is the test saying that clearTimeout should have been called?

File: src/uid2Sdk.ts (hunk 22 above)
Comment (lionell-pack-ttd) [reply src/uid2Sdk.ts:86]: It could happen, if someone switches user on a site. Or if people share a computer. I think the impact of this could be quite bad - if someone thinks they changed the user that's signed in, but they get targeted advertising for the previous user.

Remember we can't control how the publisher uses the SDK. If we give them a way to switch identity, but sometimes it puts them back to the old identity, that's a bug.

Think about a site which asks you "Is this you?" when you visit (I can't think of examples now but I see that a lot), and if you choose "no" it lets you pick from several users who sign in there sometimes. We don't want to undo the setIdentity when the refresh finishes.

File: src/integrationTests/autoRefresh.test.ts [hunk 24]
```diff
@@ -290,6 +290,60 @@ describe("when auto refreshing a non-expired identity which requires a refresh",
...
+    beforeEach(() => {
+      uid2.setIdentity(manualSetIdentity);
+      getAdvertisingTokenPromise = uid2.getAdvertisingTokenAsync();
+    });
+
+    test("should abort the refreshing request", () => {
+      expect(xhrMock.abort).toHaveBeenCalledTimes(1);
+    });
+
+    test("should reject the getAdvertisingTokenPromise", async () => {
```
Comment (lionell-pack-ttd) [inline src/integrationTests/autoRefresh.test.ts:307]: Why should it reject the promise? If something is waiting for the identity, and it's just been set, I would think the promise should be resolved with the new identity.

File: src/uid2Sdk.ts [hunk 25]
```diff
@@ -81,6 +81,7 @@ export class UID2 {
     return this.getIdentity()?.advertising_token ?? undefined;
   }
   public setIdentity(identity: Uid2Identity) {
+    this.abort(`Identity is set by setIdentity`);
```
Comment (lionell-pack-ttd) [inline src/uid2Sdk.ts:84]: I think the `abort` call is more than we need. We just need to cancel any in-progress requests.

Comment (jingyi-gao-ttd) [reply src/uid2Sdk.ts]: Thanks, I agree, we should just abort all active requests but not reject pending promise. Just updated PR

**Review by lionell-pack-ttd — APPROVED** [review]
